gitbook/graphics-transform.md

```diff
@@ -1,250 +1,228 @@
+other_box = ...
 
+if bounding_box.intersects(other_box)
+    # collision!
+end
 ```
-// the abstract base class
-class Node
-{
-public:
-
-    // ... functions to transform the node
 
-    // ... functions to manage the node's children
+The function is named `global_bounds` because it returns the bounding box of the entity in the global coordinate system, i.e. after all of its transformations (position, rotation, scale) have been applied.
 
-    void draw(sf::RenderTarget& target, const sf::Transform& parentTransform) const
-    {
-        // combine the parent transform with the node's one
-        sf::Transform combinedTransform = parentTransform * m_transform;
+There's another function that returns the bounding box of the entity in its *local* coordinate system (before its transformations are applied): `local_bounds`. This function can be used to get the initial size of an entity, for example, or to perform more specific calculations.
 
-        // let the node draw itself
-        onDraw(target, combinedTransform)
+## Object hierarchies (scene graph)
 
-        // draw its children
-        for (std::size_t i = 0; i < m_children.size(); ++i)
-            m_children[i]->draw(target, combinedTransform);
-    }
+With the custom transforms seen previously, it becomes easy to implement a hierarchy of objects in which children are transformed relative to their parent. All you have to do is pass the combined transform from parent to children when you draw them, all the way until you reach the final drawable entities (sprites, text, shapes, vertex arrays or your own drawables).
 
-private:
+```ruby
+# the abstract base class
+class Node
+    # ... functions to transform the node
 
-    virtual void onDraw(sf::RenderTarget& target, const sf::Transform& transform) const = 0;
+    # ... functions to manage the node's children
 
-    sf::Transform m_transform;
-    std::vector<Node*> m_children;
-};
+    def draw(target, parent_transform)
+        # combine the parent transform with the node's one
+        combined_transform = parent_transform * @transform;
 
-// a simple derived class: a node that draws a sprite
-class SpriteNode : public Node
-{
-public:
+        # let the node draw itself
+        on_draw(target, combined_transform)
 
-    // .. functions to define the sprite
+        # draw its children
+        @children.each do |child|
+            child.draw(target, combined_transform)
+        end
+    end
 
-private:
+    private def on_draw(target, const transform)
+        ...
+    end
+end
 
-    virtual void onDraw(sf::RenderTarget& target, const sf::Transform& transform) const
-    {
-        target.draw(m_sprite, transform);
-    }
+# a simple derived class: a node that draws a sprite
+class SpriteNode < Node
+    # .. functions to define the sprite
 
-    sf::Sprite m_sprite;
-};
+    private def on_draw(target, transform)
+        target.draw(@sprite, transform)
+    end
+end
```

Comment:
This actually doesn't involve CSFML, these are just new C++ classes. It's possible to have something similar in Crystal, but feel free to skip.